app/models/concerns/csv_importable.rb

```diff
@@ -0,0 +1,53 @@
+module CsvImportable
+  extend ActiveSupport::Concern
+
+  # rubocop:disable Metrics/BlockLength
+  class_methods do
+    def upsert_batch(batch)
+      puts "Upserting batch of #{batch.length} entries to table"
+      upsert_all(batch)
+    end
+
+    def create_from_row(row, headers_to_attributes)
+      headers_to_attributes.to_h do |header, attr_name|
+        [attr_name, attribute_types[attr_name].cast(row[header])]
+      end
+    end
+
+    def import(csv_str, batch_size)
+      rows = CSV.parse(csv_str, headers: true)
+
+      # Array of pairs mapping headers from the CSV that match a database column when converted
+      headers_to_attributes = rows.first.headers.filter_map do |h|
+        [h, convert_header(h)] if column_names.include?(convert_header(h))
+      end
+      headers_to_attributes = headers_to_attributes.nil? ? @csv_special_columns : headers_to_attributes.union(@csv_special_columns)
```

Comment:
Rather than use `@csv_special_columns` and `@csv_name_starts_with` I think it's better to define these as methods. E.g. CSV Exportable relies on `csv_attributes` which has an empty definition in the module which is then overridden in the importing classes (`def self.csv_attributes...end`)